lib/dao.js

```diff
@@ -2421,6 +2421,108 @@ DataAccessObject.updateAll = function(where, data, options, cb) {
   return cb.promise;
 };
 
+DataAccessObject.patchById = function patchById(id, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  assert(arguments.length >= 3, 'At least three arguments are required');
+
+  var Model = this;
+  var where = {};
+  var idPropertyName = idName(Model);
+
+  where[idPropertyName] = id;
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      // patchById(id, data, cb)
+      cb = options;
+      options = {};
+    }
+  }
+
+  data = data || {};
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+    'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = Model.getDataSource().connector;
+  assert(typeof connector.update === 'function',
+    'update() must be implemented by the connector');
+
+  var hookState = {};
+  this.applyProperties(data);
+
+  var context = {
+    Model: Model,
+    instanceId: id,
+    data: data,
+    hookState: hookState,
+    options: options,
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
+    doUpdate(ctx.instanceId, ctx.data);
+  });
+  function doUpdate(instanceId, data) {
+    try {
+      data = removeUndefined(data);
+      data = Model._coerce(data);
+    } catch (err) {
+      return process.nextTick(function() {
+        cb(err);
+      });
+    }
+
+    function patchCallback(err, info) {
```

Comment:
Could you please move `patchCallback` definition after `connector.update` is called? It will make the code easier to read and understand. At the moment, one has to jump a lot between different parts of this method to understand the order in which different blocks are involved.

Here is what I see now

```js
Model.notifyObserversOf('before save', // ...
// ...
Model.notifyObserversOf('loaded', // ...
// ...
Model.notifyObserversOf('after save', // ...
// ...
Model.notifyObserversOf('persist', // ...
```

And this is what I would like to see instead:

```js
Model.notifyObserversOf('before save', // ...
// ...
Model.notifyObserversOf('persist', // ...
// ...
Model.notifyObserversOf('loaded', // ...
// ...
Model.notifyObserversOf('after save', // ...
```